File: packages/sdk-java/src/main/java/xyz/tinyhumans/sdk/TinyHumansMemoryClient.java
package xyz.tinyhumans.sdk;

import xyz.tinyhumans.sdk.internal.Json;

import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.Map;
import java.util.StringJoiner;

/**
 * Client for the TinyHuman Memory API.
 * <p>
 * Requires Java 11+. Zero runtime dependencies.
 */
public class TinyHumansMemoryClient implements AutoCloseable {

    private static final String DEFAULT_BASE_URL = "https://api.tinyhumans.ai";
    private static final String DEFAULT_MODEL_ID = "neocortex-mk1";
    private static final String TINYHUMANS_BASE_URL = "TINYHUMANS_BASE_URL";

    private final String baseUrl;
    private final String token;
    private final String modelId;
    private final HttpClient httpClient;

    public TinyHumansMemoryClient(String token) {
        this(token, (String) null);
    }

    public TinyHumansMemoryClient(String token, String baseUrl) {
        this(token, null, baseUrl);
    }

    public TinyHumansMemoryClient(String token, String modelId, String baseUrl) {
        if (token == null || token.trim().isEmpty()) {
            throw new IllegalArgumentException("token is required");
        }
        this.token = token;
        this.modelId = (modelId != null && !modelId.isEmpty()) ? modelId : DEFAULT_MODEL_ID;

        String resolved = baseUrl;
        if (resolved == null || resolved.isEmpty()) {
            resolved = System.getenv(TINYHUMANS_BASE_URL);
        }
        if (resolved == null || resolved.isEmpty()) {
            resolved = DEFAULT_BASE_URL;
        }
        this.baseUrl = resolved.replaceAll("/+$", "");

        this.httpClient = HttpClient.newBuilder()
                .connectTimeout(Duration.ofSeconds(30))
                .build();
    }

    /** Insert (ingest) a document into memory. POST /memory/insert */
    public InsertMemoryResponse insertMemory(InsertMemoryParams params) {
        Map<String, Object> body = params.toMap();
        Map<String, Object> response = post("/memory/insert", body);
        return InsertMemoryResponse.fromMap(response);
    }

    /** Recall context from Master node. POST /memory/recall */
    public RecallMemoryResponse recallMemory(RecallMemoryParams params) {
        Map<String, Object> body = params.toMap();
        Map<String, Object> response = post("/memory/recall", body);
        return RecallMemoryResponse.fromMap(response);
    }

    /** Delete memory (admin). POST /memory/admin/delete */
    public DeleteMemoryResponse deleteMemory(DeleteMemoryParams params) {
        Map<String, Object> body = params.toMap();
        Map<String, Object> response = post("/memory/admin/delete", body);
        return DeleteMemoryResponse.fromMap(response);
    }

    /** Query memory via RAG. POST /memory/query */
    public QueryMemoryResponse queryMemory(QueryMemoryParams params) {
        Map<String, Object> body = params.toMap();
        Map<String, Object> response = post("/memory/query", body);
        return QueryMemoryResponse.fromMap(response);
    }

    /** Recall memories from Ebbinghaus bank. POST /memory/memories/recall */
    public RecallMemoriesResponse recallMemories(RecallMemoriesParams params) {
        Map<String, Object> body = params.toMap();
        Map<String, Object> response = post("/memory/memories/recall", body);
        return RecallMemoriesResponse.fromMap(response);
    }

    /** Insert a document. POST /memory/documents */
    public Map<String, Object> insertDocument(InsertDocumentParams params) {
        return post("/memory/documents", params.toMap());
    }

    /** Insert documents in batch. POST /memory/documents/batch */
    public Map<String, Object> insertDocumentsBatch(InsertDocumentsBatchParams params) {
        return post("/memory/documents/batch", params.toMap());
    }

    /** List documents. GET /memory/documents */
    public Map<String, Object> listDocuments(ListDocumentsParams params) {
        return sendGet("/memory/documents", params != null ? params.toQueryParams() : null);
    }

    /** Get a document by ID. GET /memory/documents/{documentId} */
    public Map<String, Object> getDocument(GetDocumentParams params) {
        params.validate();
        return sendGet("/memory/documents/" + params.getDocumentId(), params.toQueryParams());
    }

    /** Delete a document. DELETE /memory/documents/{documentId} */
    public Map<String, Object> deleteDocument(String documentId, String namespace) {
        if (documentId == null || documentId.isEmpty()) {
            throw new IllegalArgumentException("documentId is required");
        }
        if (namespace == null || namespace.isEmpty()) {
            throw new IllegalArgumentException("namespace is required");
        }
        return sendDelete("/memory/documents/" + documentId, Map.of("namespace", namespace));
    }

    /** Generate reflective thoughts. POST /memory/memories/thoughts */
    public Map<String, Object> recallThoughts(RecallThoughtsParams params) {
        return post("/memory/memories/thoughts", params != null ? params.toMap() : Map.of());
    }

    /** Query memory context. POST /memory/queries */
    public Map<String, Object> queryMemoryContext(QueryMemoryContextParams params) {
        return post("/memory/queries", params.toMap());
    }

    /** Record entity interactions. POST /memory/interact */
    public Map<String, Object> interactMemory(InteractMemoryParams params) {
        return post("/memory/interact", params.toMap());
    }

    /** Record interaction signals. POST /memory/interactions */
    public Map<String, Object> recordInteractions(InteractMemoryParams params) {
        return post("/memory/interactions", params.toMap());
    }

    /** Chat with DeltaNet memory cache. POST /memory/chat */
    public Map<String, Object> chatMemory(ChatMemoryParams params) {
        return post("/memory/chat", params.toMap());
    }

    /** Chat with memory context. POST /memory/conversations */
    public Map<String, Object> chatMemoryContext(ChatMemoryParams params) {
        return post("/memory/conversations", params.toMap());
    }

    @Override
    public void close() {
        // HttpClient does not require explicit close in Java 11
    }

    // ---- Internal ----

    private Map<String, Object> post(String path, Map<String, Object> body) {
        String jsonBody = Json.serialize(body);
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(baseUrl + path))
                .header("Content-Type", "application/json")
                .header("Authorization", "Bearer " + token)
                .header("X-Model-Id", modelId)
                .POST(HttpRequest.BodyPublishers.ofString(jsonBody))
                .timeout(Duration.ofSeconds(30))
                .build();

        HttpResponse<String> response;
        try {
            response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        } catch (Exception e) {
            throw new RuntimeException("HTTP request failed: " + e.getMessage(), e);
        }

        return handleResponse(response);
    }

    Map<String, Object> sendGet(String path, Map<String, String> params) {
        String url = baseUrl + path + buildQueryString(params);
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .header("Authorization", "Bearer " + token)
                .header("X-Model-Id", modelId)
                .GET()
                .timeout(Duration.ofSeconds(30))
                .build();

        HttpResponse<String> response;
        try {
            response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        } catch (Exception e) {
            throw new RuntimeException("HTTP request failed: " + e.getMessage(), e);
        }

        return handleResponse(response);
    }

    Map<String, Object> sendDelete(String path, Map<String, String> params) {
        String url = baseUrl + path + buildQueryString(params);
        HttpRequest request = HttpRequest.newBuilder()
                .uri(URI.create(url))
                .header("Authorization", "Bearer " + token)
                .header("X-Model-Id", modelId)
                .DELETE()
                .timeout(Duration.ofSeconds(30))
                .build();

        HttpResponse<String> response;
        try {
            response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        } catch (Exception e) {
            throw new RuntimeException("HTTP request failed: " + e.getMessage(), e);
        }

        return handleResponse(response);
    }

    private String buildQueryString(Map<String, String> params) {
        if (params == null || params.isEmpty()) return "";
        StringJoiner joiner = new StringJoiner("&", "?", "");
        for (Map.Entry<String, String> entry : params.entrySet()) {
            joiner.add(URLEncoder.encode(entry.getKey(), StandardCharsets.UTF_8)
                    + "=" + URLEncoder.encode(entry.getValue(), StandardCharsets.UTF_8));
        }
        return joiner.toString();
    }

    private Map<String, Object> handleResponse(HttpResponse<String> response) {
        String text = response.body();
        Map<String, Object> json;
        try {
            json = (text != null && !text.isEmpty()) ? Json.parse(text) : Map.of();
        } catch (Exception e) {
            throw new TinyHumansError(
                    "HTTP " + response.statusCode() + ": non-JSON response",
                    response.statusCode(),
                    text
            );
        }

        if (response.statusCode() < 200 || response.statusCode() >= 300) {
            String message = "HTTP " + response.statusCode();
            Object error = json.get("error");
            if (error instanceof String) {
                message = (String) error;
            }
            throw new TinyHumansError(message, response.statusCode(), json);
        }

        return json;
    }
}

File: packages/sdk-dart/lib/src/tinyhumans_memory_client.dart
import 'dart:convert';
import 'dart:io' show Platform;

import 'package:http/http.dart' as http;

import 'tinyhumans_error.dart';
import 'types.dart';

class TinyHumansMemoryClient {
  static const _defaultBaseUrl = 'https://api.tinyhumans.ai';
  static const _defaultModelId = 'neocortex-mk1';

  final String _token;
  final String _baseUrl;
  final String _modelId;
  final http.Client _httpClient;
  final bool _ownsClient;

  TinyHumansMemoryClient(
    String token, {
    String? modelId,
    String? baseUrl,
    http.Client? httpClient,
  })  : _token = token,
        _modelId = (modelId != null && modelId.trim().isNotEmpty)
            ? modelId
            : _defaultModelId,
        _baseUrl = (baseUrl ??
                Platform.environment['TINYHUMANS_BASE_URL'] ??
                _defaultBaseUrl)
            .replaceAll(RegExp(r'/+$'), ''),
        _httpClient = httpClient ?? http.Client(),
        _ownsClient = httpClient == null {
    if (token.trim().isEmpty) {
      throw ArgumentError('token is required');
    }
  }

  Future<InsertMemoryResponse> insertMemory(InsertMemoryParams params) async {
    params.validate();
    final result = await _post('/memory/insert', params.toJson());
    return InsertMemoryResponse.fromJson(result);
  }

  Future<RecallMemoryResponse> recallMemory([RecallMemoryParams? params]) async {
    params ??= RecallMemoryParams();
    params.validate();
    final result = await _post('/memory/recall', params.toJson());
    return RecallMemoryResponse.fromJson(result);
  }

  Future<DeleteMemoryResponse> deleteMemory([DeleteMemoryParams? params]) async {
    params ??= DeleteMemoryParams();
    params.validate();
    final result = await _post('/memory/admin/delete', params.toJson());
    return DeleteMemoryResponse.fromJson(result);
  }

  Future<QueryMemoryResponse> queryMemory(QueryMemoryParams params) async {
    params.validate();
    final result = await _post('/memory/query', params.toJson());
    return QueryMemoryResponse.fromJson(result);
  }

  Future<RecallMemoriesResponse> recallMemories(
      [RecallMemoriesParams? params]) async {
    params ??= RecallMemoriesParams();
    params.validate();
    final result = await _post('/memory/memories/recall', params.toJson());
    return RecallMemoriesResponse.fromJson(result);
  }

  Future<Map<String, dynamic>> _post(
      String path, Map<String, dynamic> body) async {
    final url = Uri.parse('$_baseUrl$path');
    final response = await _httpClient.post(
      url,
      headers: {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer $_token',
        'X-Model-Id': _modelId,
      },
      body: jsonEncode(body),
    );
    return _handleResponse(response);
  }

  Future<Map<String, dynamic>> _get(String path,
      [Map<String, String>? queryParams]) async {
    var uri = Uri.parse('$_baseUrl$path');
    if (queryParams != null && queryParams.isNotEmpty) {
      uri = uri.replace(queryParameters: queryParams);
    }
    final response = await _httpClient.get(
      uri,
      headers: {
        'Authorization': 'Bearer $_token',
        'X-Model-Id': _modelId,
      },
    );
    return _handleResponse(response);
  }

  Future<Map<String, dynamic>> _delete(String path,
      [Map<String, String>? queryParams]) async {
    var uri = Uri.parse('$_baseUrl$path');
    if (queryParams != null && queryParams.isNotEmpty) {
      uri = uri.replace(queryParameters: queryParams);
    }
    final response = await _httpClient.delete(
      uri,
      headers: {
        'Authorization': 'Bearer $_token',
        'X-Model-Id': _modelId,
      },
    );
    return _handleResponse(response);
  }

  Map<String, dynamic> _handleResponse(http.Response response) {
    Map<String, dynamic> json;
    try {
      json = response.body.isNotEmpty
          ? jsonDecode(response.body) as Map<String, dynamic>
          : <String, dynamic>{};
    } catch (_) {
      throw TinyHumansError(
        'HTTP ${response.statusCode}: non-JSON response',
        response.statusCode,
        response.body,
      );
    }

    if (response.statusCode < 200 || response.statusCode >= 300) {
      final message =
          json['error'] as String? ?? 'HTTP ${response.statusCode}';
      throw TinyHumansError(message, response.statusCode, response.body);
    }

    return json;
  }

  void close() {
    if (_ownsClient) {
      _httpClient.close();
    }
  }
}
